rllib/connectors/tests/test_agent.py

```diff
@@ -274,105 +276,85 @@ def test_vr_connector_shift_by_one(self):
         }
 
         obs_arrs = np.arange(10)[:, None] + 1
-        ctx = ConnectorContext(view_requirements=view_rq_dict)
+        config = PPOConfig().to_dict()
+        ctx = ConnectorContext(
+            view_requirements=view_rq_dict, config=config, is_policy_recurrent=True
+        )
         c = ViewRequirementAgentConnector(ctx)
 
-        for is_training in [True, False]:
-            c.is_training(is_training)
-            for i, obs_arr in enumerate(obs_arrs):
-                data = AgentConnectorDataType(0, 1, dict(obs=obs_arr))
-                processed = c([data])
-                for_action = processed[0].data.for_action
-
-                self.assertTrue("next_state" not in for_action)
-                check(for_action["state"], obs_arrs[i])
-                if i == 0:
-                    check(for_action["prev_state"], np.array([0]))
-                else:
-                    check(for_action["prev_state"], obs_arrs[i - 1])
+        # keep a running list of observations
+        obs_list = []
+        for t, obs in enumerate(obs_arrs):
+            # t=0 is the next state of t=-1
+            data = AgentConnectorDataType(
+                0, 1, {SampleBatch.NEXT_OBS: obs, SampleBatch.T: t - 1}
+            )
+            processed = c([data])  # env.reset() for t == -1 else env.step()
+            for_action = processed[0].data.for_action
+            # add cur obs to the list
+            obs_list.append(obs)
 
-    def test_vr_connector_causal_slice(self):
-        """Test that the ViewRequirementConnector can handle slice shifts correctly.
+            if t == 0:
+                check(for_action["prev_state"], for_action["state"])
+            else:
+                # prev state should be equal to the prev time step obs
+                check(for_action["prev_state"], obs_list[-2][None])
```

Comment:
curious, what is this None key?

Reply:
to add a single batch to the front of it. i.e if `obs_list[-2].shape = (1,)` then `obs_list[-2][None].shape = (1,1)`

Reply:
wow, didn't know you can do that. 👍